### PR TITLE
PERF-#6433: Implement '.dropna()' using map-reduce pattern

### LINKED ISSUE #6433: Reimplement `.dropna()` using MapReduce pattern
Currently, it's implemented as a full-axis function, which works terribly slow in case we want to perform dropna along columns.

We can simply rewrite `.dropna()` using MapReduce pattern, which should significantly improve performance.

It's proposed to write something like this:
```python
# pandas query compiler
def dropna(self, **kwargs):
    masks = self._modin_frame.map(
        func=lambda df: df.isna().all()
    )

    def func(df, *masks):
        non_nan_cols = conjunction_masks(masks)                
        return df[non_nan_cols]
    
    result = self._modin_frame.map_with_broadcast(func, masks)
    return self.__constructor__(result)
```

### PR DESCRIPTION
**Pull request by dchigarev**
This PR replaces a full-column implementation of `.dropna()` with a chain of two row-axis/map functions. 

At the first stage, we submit a row-axis function evaluating whether each column should be dropped for this specific row partition and builds a boolean mask for that matter, at the second stage those masks are being merged and the columns to be dropped are being dropped.

<details><summary>perf results</summary>

Here's the comparison for `df.dropna(axis=1, how="any")` between these two implementations:

<details><summary>code to measure</summary>

```python
import modin.pandas as pd
import modin.config as cfg
import numpy as np
from timeit import default_timer as timer

cfg.BenchmarkMode.put(True)

# start all the workers
pd.DataFrame([np.arange(cfg.MinPartitionSize.get()) for _ in range(cfg.NPartitions.get() ** 2)]).to_numpy()

shapes = [
    [5000, 5000],
    [1_000_000, 32],
    [10_000_000, 32],
    [10_000_000, 256]
]

for how in ("any",):
    for shape in shapes:
        # data-gen was copied from asv TimeDropna benchmark
        row, col = shape
        df = pd.DataFrame(np.random.randn(row, col))
        df.iloc[row // 20 : row // 10, col // 3 : col // 2] = np.nan
        df["foo"] = "bar"

        t1 = timer()
        df.dropna(axis=1, how=how)
        print(f"{shape=}, {how=}: {timer() - t1}")
```

</details>

image

</details>

## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6433 
- [x] tests are passing
- [x] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 1]
```diff
@@ -2868,13 +2909,19 @@ def _pick_axis(get_axis, sizes_cache):
         if axis == 0:
             # Pass shape caches instead of values in order to not trigger shape computation.
             new_index, new_row_lengths = _pick_axis(
-                self._get_index, self._row_lengths_cache
+                self.copy_index_cache, self._row_lengths_cache
```
Comment (dchigarev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:2912]: `.copy_index_cache()` doesn't trigger index materialization in comparison with `._get_index()` that does it

File: modin/core/dataframe/pandas/dataframe/dataframe.py [hunk 2]
```diff
@@ -2829,6 +2852,13 @@ def broadcast_apply(
             Modin DataFrame to broadcast.
         join_type : str, default: "left"
             Type of join to apply.
+        copartition : bool, default: True
```
Comment (anmyachev) [inline modin/core/dataframe/pandas/dataframe/dataframe.py:2855]: The introduction of this parameter looks like a fix for situations where the function was called on unaligned partitions. Let's separate this into a separate pull request and add tests.

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 3]
```diff
@@ -2828,6 +2828,62 @@ def setitem_builder(df, internal_indices=[]):  # pragma: no cover
...
+        no_thresh_passed = kwargs.get("thresh", no_default) in (
+            no_default,
+            None,
+        )
+        # FIXME: this is a naive workaround for this problem: https://github.com/modin-project/modin/issues/5394
+        # if there are too many partitions then all non-full-axis implementations start acting very badly.
+        # The here threshold is pretty random though it works fine on simple scenarios
+        processable_amount_of_partitions = (
+            np.prod(self._modin_frame._partitions.shape) < CpuCount.get() * 32
+        )
```
Comment (anmyachev) [inline modin/core/storage_formats/pandas/query_compiler.py:2841]: Let's not mix abstraction levels and separate this condition into a dataframe method.

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 4]
```diff
@@ -2828,6 +2828,62 @@ def setitem_builder(df, internal_indices=[]):  # pragma: no cover
...
+
+        if is_column_wise and no_thresh_passed and processable_amount_of_partitions:
+            how = kwargs.get("how", "any")
+            subset = kwargs.get("subset")
+            how = "any" if how in (no_default, None) else how
+            condition = lambda df: getattr(df, how)()  # noqa: E731 (lambda assignment)
+
+            def mapper(df: pandas.DataFrame):
```
Comment (anmyachev) [inline modin/core/storage_formats/pandas/query_compiler.py:2849]: A little comment on what this kernel is for is useful, since the name of this function is generic.

File: modin/core/storage_formats/pandas/query_compiler.py [hunk 5]
```diff
@@ -2828,6 +2828,62 @@ def setitem_builder(df, internal_indices=[]):  # pragma: no cover
...
+                    mask.update(subset_mask)
+                else:
+                    mask = condition(df.isna())
+                return mask.to_frame().T
+
+            masks = self._modin_frame.apply_full_axis(
+                func=mapper, axis=1, keep_partitioning=True
+            )
+
+            def reduce(df: pandas.DataFrame, mask: pandas.DataFrame):
```
Comment (anmyachev) [inline modin/core/storage_formats/pandas/query_compiler.py:2866]: A little comment on what this kernel is for is useful, since the name of this function is generic.

**Review by anmyachev — CHANGES_REQUESTED** [review]

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 3 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:2841]: > Let's separate this condition into a dataframe method.

well, that would be quite hard...

here I want to dispatch between two different implementations of `.dropna()`. Since dataframe level doesn't define `dropna`, then I have to do this dispatching at the QC level.

The idea of a "general dispatcher" at the dataframe level was discussed at #5394 but trying suggested ideas didn't really find success, so there's a lot of work to be done in this regard

File: modin/core/dataframe/pandas/dataframe/dataframe.py (hunk 2 above)
Comment (dchigarev) [reply modin/core/dataframe/pandas/dataframe/dataframe.py:2855]: It's not the fix you're talking about.

If you'll look at the previous implementation of this method, you'll see that the default behavior was to perform co-partitioning unconditionally. This parameter just gives an ability to provide a hint on whether we can safely skip the co-partitioning if we're sure that the frames are identically partitioned.

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 4 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:2849]: done

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 5 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py:2871]: done

File: modin/core/storage_formats/pandas/query_compiler.py (hunk 3 above)
Comment (dchigarev) [reply modin/core/storage_formats/pandas/query_compiler.py]: Except for the `.processable_amount_of_partitions_for_dropna` field, I simply added a property `.num_parts` and kept the deciding logic at the QC, so the line now looks like:
```python
processable_amount_of_partitions = self._modin_frame.num_parts < CpuCount.get() * 32
```
It feels wrong adding the `.processable_amount_of_partitions_for_dropna` field, as the `modin_frame` layer should know nothing about the `dropna` and its implementation desires in terms of partitions amount

**Review by anmyachev — APPROVED** [review]
LGTM!

**Review by YarShev — APPROVED** [review]
